scripthost/src/main/kotlin/com/commonwealthrobotics/bowlerkernel/scripthost/Session.kt

```diff
@@ -0,0 +1,271 @@
+     */
+    private suspend fun runRequest(
+        runRequest: RunRequest,
+        sessionChannel: SendChannel<SessionServerMessage>
+    ): Either<Throwable, Any?> {
+        val scriptLoader = getKoin().get<ScriptLoader>()
+
+        val script = sessionChannel.withTask(
+            runRequest.requestId,
+            nextTaskID.getAndIncrement(),
+            "Initializing ${runRequest.file.path}"
+        ) {
+            val script = scriptLoader.resolveAndLoad(runRequest.file, runRequest.devsList, runRequest.environmentMap)
+            scriptRequestMap[runRequest.requestId] = script
+            script
+        }
+
+        return sessionChannel.withTask(
+            runRequest.requestId,
+            nextTaskID.getAndIncrement(),
+            "Running ${runRequest.file.path}"
+        ) {
+            script.map {
+                it.start(emptyList(), null)
+                val scriptResult = it.join()
+                logger.info { "Script returned:\n$scriptResult" }
+                when (scriptResult) {
+                    // Throw when the script errored so that the task and request fail
+                    is Either.Left -> throw scriptResult.a
+                    is Either.Right -> scriptResult.b
+                }
+            }
+        }
+    }
+
+    override suspend fun getCredentialsFor(remote: String): Credentials {
+        val msg = requests.call(
+            SessionServerMessage.newBuilder().apply {
+                credentialsRequestBuilder.remote = remote
+            }
+        )
+
+        return when {
+            msg.hasCredentialsResponse() -> {
+                val res = msg.credentialsResponse
+                when {
+                    res.hasBasic() -> Credentials.Basic(res.basic.username, res.basic.password)
+                    res.hasOauth() -> Credentials.OAuth(res.oauth.token)
+                    else -> Credentials.Anonymous
+                }
+            }
+            msg.hasError() -> throw IllegalStateException("Credentials request error: " + msg.error.description)
+            else -> throw IllegalStateException("Unknown credentials response: $msg")
+        }
+    }
+
+    override suspend fun getTwoFactorFor(remote: String): String {
+        val msg = requests.call(
+            SessionServerMessage.newBuilder().apply {
+                twoFactorRequestBuilder.description = remote
+            }
+        )
+
+        return when {
+            msg.hasTwoFactorResponse() -> msg.twoFactorResponse.twoFactor
+            msg.hasError() -> throw IllegalStateException("2FA request error: " + msg.error.description)
+            else -> throw IllegalStateException("Unknown 2FA response: $msg")
+        }
+    }
+
+    /**
+     * @return true if the [builder] is a "request" type (meaning it has a request ID).
+     */
+    private fun isRequest(builder: GeneratedMessageV3.Builder<*>) =
+        (builder as SessionServerMessage.Builder).let {
```

Comment:
Unsafe cast feels like the wrong approach here. As per above I think that the canonical builder type should be the concrete one in the first place.

Reply:
Resolved because I fixed the parameter type.